File: java/leetcode/PostorderTraversal.java
//Leetcode problem 145 Binary Tree PostorderTraversal
//Solution written by Xuqiang Fang on 25 March 2018 

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedList;

class TreeNode{
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x){val = x;}
}


class Solution{
	public List<Integer> recursive(TreeNode root){
		List<Integer> list = new ArrayList<>();
		if(root == null)
			return list;

		list.addAll(recursive(root.left));
		list.addAll(recursive(root.right));
		list.add(root.val);
		return list;
	}	
	public List<Integer> postorderTraversal(TreeNode root){
		List<Integer> list = new ArrayList<>();
		LinkedList<TreeNode> stack = new LinkedList<>();
		TreeNode current = root;
		TreeNode pre = null;

		while(current != null || !stack.isEmpty()){
			while(current != null){
				stack.push(current);
				current = current.left;
			}

			current = stack.pop();
			if(current.right != null && pre != current.right){
				stack.push(current);
				current = current.right;
				continue;
			}
			list.add(current.val);
			pre = current;
			current = null;
		}
		return list;
	}
}

public class PostorderTraversal{
	public static void main(String[] args){
		Solution s = new Solution();
		TreeNode root = new TreeNode(1);
		TreeNode a = new TreeNode(2);
		TreeNode b = new TreeNode(3);
		TreeNode c = new TreeNode(8);
		TreeNode d = new TreeNode(9);
		TreeNode e = new TreeNode(5);
		root.left = c;
		c.left = d;
		c.right = e;

		root.right = a;
		a.left = b;
		System.out.println(s.recursive(root));
		System.out.println(s.postorderTraversal(root));
	}
}

File: python/leetcode/PostorderTraversal.py
'''
Leetcode problem No 145 Binary Tree Postorder Traversal
Solution written by Xuqiang Fang on 20 July, 2018
'''

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans, stack = [], [(root, False)]
        while stack:
            root, visited = stack.pop()
            if not root:
                continue
            if visited:
                ans.append(root.val)
            else:
                stack.append((root.right, False))
                stack.append((root.left, False))
                stack.append((root, True))
        return ans

def main():
    s = Solution()
    r = TreeNode(1)
    a = TreeNode(2)
    b = TreeNode(3)
    c = TreeNode(4)
    d = TreeNode(5)
    r.left = a
    r.right = b
    a.left = c
    a.right = d
    ans = s.postorderTraversal(r)
    for i in ans:
        print(i)
if __name__ == '__main__':
    main()
